### PR TITLE
feat: Added support for memo

### LINKED ISSUE #88: [Task] Implement support for Memo in transactions
# Task

Contributors are able to complete this task and earn thenewboston coins.  Check out the labels to learn how much you can earn for contributing by completing this task.  Please make sure to be honest if you wish to contribute by saying you can't finish this and we can just un-assign you with no harm done!  There is no point in delaying tasks from being completed for miscommunication!

## Overview
Based on preliminary information available here https://github.com/thenewboston-developers/Management/issues/81 and implementation in account manager https://github.com/thenewboston-developers/Account-Manager/pull/615
Provide an implementation for the Memo feature for transaction on the chain

## Behavior

Users should be able to use the SDK to create/perform transactions that include a memo/description of the transaction.

an example implementation can be found here https://github.com/thenewboston-developers/Account-Manager/pull/615

### PR DESCRIPTION
**Pull request by AbhayAysola**
Closes #88 
Account Number: 8928526805de48e4bf2ed2a9b4b839e6b2603018ecdfbf4cffdb2065e01a3ed1

### CODE REVIEW

**Comment by AbhayAysola** [comment]
@zinoadidi This should work as far as I know, but since I can't test it anywhere it might not work as expected.

**Comment by zinoadidi** [comment]
will you update when you manage to test it;
also, tests should be updated as well to reflect possibility with memo
and docs
and I saw some places where 'empty string' was used instead of STRING  type didn't pay much attention but could be something to check in code

**Comment by AbhayAysola** [comment]
> will you update when you manage to test it;
> also, tests should be updated as well to reflect possibility with memo
> and docs
> and I saw some places where 'empty string' was used instead of STRING type didn't pay much attention but could be something to check in code

Sure, I will update docs and tests when I can confirm that it works. 
The empty string is done on purpose since memo isn't mandatory.

**Comment by zinoadidi** [comment]
ok cool; 
Maybe something to check, I remember with TS it doesn't include the file if its not set
as I read in the discord chat, bucky says its possible to omit the field from request if user doesn't provide it.

Its just a theory, we can check it

**Comment by AbhayAysola** [comment]
Yeah that shouldn't be an issue we can test it when the banks are up.

File: src/utils/transfer-details.ts
Comment (zinoadidi) [inline src/utils/transfer-details.ts:14]: .matches(/^[a-zA-Z0-9_ ]*$/, 'Memo can only contain alphanumeric characters, spaces, and underscores')
        .max(64, 'Memo cannot exceed 64 characters'),
(memo: string): string => memo.trim();


What do you think some of these validations are maybe not needed?
clients been clients, we cant trust them to follow all requirements without enforcing it

Comment (AbhayAysola) [reply src/utils/transfer-details.ts:14]: Well I can add that in if you'd like, but the server should respond with an error if those requirements aren't met. It's your call

**Comment by tomijaga** [comment]
Hey @AbhayAysola there is already a PR on the account manager for the memo. 
I think you can use this to implement the memo here since it was made directly from the core team

**Comment by tomijaga** [comment]
Here's the PR https://github.com/thenewboston-developers/Account-Manager/pull/615

**Comment by AbhayAysola** [comment]
> Here's the PR [thenewboston-developers/Account-Manager#615](https://github.com/thenewboston-developers/Account-Manager/pull/615)

Yep that's exactly what I did

**Comment by AbhayAysola** [comment]
> @AbhayAysola Okay so all that's left is:
> 
> * check that the memo only has alphanumeric characters by using the
>   same  regex expression as in the PR:   `/^[a-zA-Z0-9_ ]*$/`
> * Remove leading spaces using the `.trim()` method
> * And if memo is empty just remove it from the transaction object
> 
> Do all of these in the `createTransaction()` method in the **PaymentHandler** class
> 
> This is the last thing needed for the migration to the account manager. I could make the updates on your branch if you want

I added those in

**Comment by tomijaga** [comment]
@AbhayAysola Nice! 
@zinoadidi We can merge this in now and create new tasks for test and docs

**Comment by AbhayAysola** [comment]
I tested it and made some changes @zinoadidi @tomijaga 

Should I update tests and docs in this PR or will it be in another task?

**Comment by zinoadidi** [comment]
@AbhayAysola you can go ahead and update test and docs as well; mention me when its ready and we can have this feature released as well;

Good job!

**Comment by AbhayAysola** [comment]
@zinoadidi I updated the docs but there wasn't anything to change in the tests since it is just an addition of a field.

**Comment by zinoadidi** [comment]
@AbhayAysola seen.
I do think the field should be reflected in the tests and transaction that don't meet the requirements should be included in test / updated to work with and reject memo as well
Else it will be possible to remote the field unknowingly in future and the tests will not catch it

**Comment by AbhayAysola** [comment]
> @AbhayAysola seen.
> I do think the field should be reflected in the tests and transaction that don't meet the requirements should be included in test / updated to work with and reject memo as well
> Else it will be possible to remote the field unknowingly in future and the tests will not catch it

Hmm I guess you're right. I'll add that in

File: src/payment-handler.ts
Comment (zinoadidi) [inline src/payment-handler.ts:57]: do we want to log to console in sdk?

Comment (AbhayAysola) [reply src/payment-handler.ts:57]: Probably not, that's my bad

**Review by zinoadidi — APPROVED** [review]

**Comment by zinoadidi** [comment]
@AbhayAysola confirm your account number

**Comment by AbhayAysola** [comment]
> @AbhayAysola confirm your account number

8928526805de48e4bf2ed2a9b4b839e6b2603018ecdfbf4cffdb2065e01a3ed1

**Pull request merged**
